File: go/internal/split/split.go
package split

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// ParseSize parses a human-readable size (see SPEC.md section 5).
func ParseSize(s string) (int64, error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, errors.New("empty size")
	}
	up := strings.ToUpper(s)
	type unit struct {
		suffix string
		mult   int64
	}
	// Order matters: check longer suffixes first.
	units := []unit{
		{"KIB", 1 << 10}, {"MIB", 1 << 20}, {"GIB", 1 << 30}, {"TIB", 1 << 40},
		{"KB", 1e3}, {"MB", 1e6}, {"GB", 1e9}, {"TB", 1e12},
		{"K", 1 << 10}, {"M", 1 << 20}, {"G", 1 << 30}, {"T", 1 << 40},
		{"B", 1},
	}
	for _, u := range units {
		if strings.HasSuffix(up, u.suffix) {
			num := strings.TrimSpace(up[:len(up)-len(u.suffix)])
			v, err := strconv.ParseInt(num, 10, 64)
			if err != nil {
				return 0, fmt.Errorf("invalid size %q: %w", s, err)
			}
			if v < 0 {
				return 0, fmt.Errorf("invalid size %q: negative", s)
			}
			return v * u.mult, nil
		}
	}
	v, err := strconv.ParseInt(up, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("invalid size %q: %w", s, err)
	}
	return v, nil
}

// Writer splits everything written to it into sequential parts named
// "<base>.NNN" (NNN zero-padded to 3 digits), each at most size bytes.
type Writer struct {
	base    string
	size    int64
	part    int
	cur     *os.File
	written int64
	parts   []string
}

func NewWriter(base string, size int64) *Writer {
	return &Writer{base: base, size: size}
}

func (w *Writer) Write(p []byte) (int, error) {
	total := 0
	for len(p) > 0 {
		if w.cur == nil || w.written >= w.size {
			if err := w.rotate(); err != nil {
				return total, err
			}
		}
		space := w.size - w.written
		take := int64(len(p))
		if take > space {
			take = space
		}
		n, err := w.cur.Write(p[:take])
		w.written += int64(n)
		total += n
		p = p[n:]
		if err != nil {
			return total, err
		}
	}
	return total, nil
}

func (w *Writer) rotate() error {
	if w.cur != nil {
		if err := w.cur.Close(); err != nil {
			return err
		}
	}
	w.part++
	name := fmt.Sprintf("%s.%03d", w.base, w.part)
	f, err := os.Create(name)
	if err != nil {
		return err
	}
	w.cur = f
	w.written = 0
	w.parts = append(w.parts, name)
	return nil
}

// Close flushes the final part. An input with no bytes still yields one empty part.
func (w *Writer) Close() error {
	if w.cur == nil {
		if err := w.rotate(); err != nil {
			return err
		}
	}
	return w.cur.Close()
}

// Parts returns the list of files written so far.
func (w *Writer) Parts() []string { return w.parts }

var suffixRe = regexp.MustCompile(`^(.*)\.(\d+)$`)

// Stem strips a trailing ".NNN" part suffix, if present.
func Stem(arg string) string {
	if m := suffixRe.FindStringSubmatch(arg); m != nil {
		return m[1]
	}
	return arg
}

// FindParts locates and orders the part files for a given part path or stem.
func FindParts(arg string) ([]string, error) {
	stem := arg
	if m := suffixRe.FindStringSubmatch(arg); m != nil {
		stem = m[1]
	}
	dir := filepath.Dir(stem)
	base := filepath.Base(stem)
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	partRe := regexp.MustCompile("^" + regexp.QuoteMeta(base) + `\.(\d+)$`)
	type pp struct {
		n    int
		path string
	}
	var found []pp
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		if m := partRe.FindStringSubmatch(e.Name()); m != nil {
			num, _ := strconv.Atoi(m[1])
			found = append(found, pp{num, filepath.Join(dir, e.Name())})
		}
	}
	if len(found) == 0 {
		return nil, fmt.Errorf("no parts found for %q", arg)
	}
	sort.Slice(found, func(i, j int) bool { return found[i].n < found[j].n })
	paths := make([]string, len(found))
	for i, f := range found {
		paths[i] = f.path
	}
	return paths, nil
}

type multiReadCloser struct {
	io.Reader
	closers []io.Closer
}

func (m *multiReadCloser) Close() error {
	var firstErr error
	for _, c := range m.closers {
		if err := c.Close(); err != nil && firstErr == nil {
			firstErr = err
		}
	}
	return firstErr
}

// OpenParts returns a reader that streams the concatenation of the parts.
func OpenParts(parts []string) (io.ReadCloser, error) {
	readers := make([]io.Reader, 0, len(parts))
	closers := make([]io.Closer, 0, len(parts))
	for _, p := range parts {
		f, err := os.Open(p)
		if err != nil {
			for _, c := range closers {
				c.Close()
			}
			return nil, err
		}
		readers = append(readers, f)
		closers = append(closers, f)
	}
	return &multiReadCloser{Reader: io.MultiReader(readers...), closers: closers}, nil
}

File: py/packer/split.py
"""Split/merge layer (see SPEC.md 5)."""

import os
import re

_SUFFIX_RE = re.compile(r"^(.*)\.(\d+)$")

_UNITS = [
    ("KIB", 1 << 10), ("MIB", 1 << 20), ("GIB", 1 << 30), ("TIB", 1 << 40),
    ("KB", 10 ** 3), ("MB", 10 ** 6), ("GB", 10 ** 9), ("TB", 10 ** 12),
    ("K", 1 << 10), ("M", 1 << 20), ("G", 1 << 30), ("T", 1 << 40),
    ("B", 1),
]


def parse_size(s):
    s = s.strip()
    if not s:
        raise ValueError("empty size")
    up = s.upper()
    for suffix, mult in _UNITS:
        if up.endswith(suffix):
            num = up[: len(up) - len(suffix)].strip()
            v = int(num)
            if v < 0:
                raise ValueError("negative size")
            return v * mult
    return int(up)


class SplitWriter:
    """A binary writer that splits into '<base>.NNN' parts of at most size bytes."""

    def __init__(self, base, size):
        self.base = base
        self.size = size
        self.part = 0
        self.cur = None
        self.written = 0
        self.parts = []

    def write(self, data):
        mv = memoryview(bytes(data))
        total = 0
        while len(mv):
            if self.cur is None or self.written >= self.size:
                self._rotate()
            space = self.size - self.written
            take = min(len(mv), space)
            self.cur.write(mv[:take])
            self.written += take
            total += take
            mv = mv[take:]
        return total

    def _rotate(self):
        if self.cur is not None:
            self.cur.close()
        self.part += 1
        name = "%s.%03d" % (self.base, self.part)
        self.cur = open(name, "wb")
        self.written = 0
        self.parts.append(name)

    def flush(self):
        if self.cur is not None:
            self.cur.flush()

    def close(self):
        if self.cur is None:
            self._rotate()
        if self.cur is not None:
            self.cur.close()
            self.cur = None


def stem(arg):
    m = _SUFFIX_RE.match(arg)
    return m.group(1) if m else arg


def find_parts(arg):
    st = stem(arg)
    directory = os.path.dirname(st) or "."
    base = os.path.basename(st)
    part_re = re.compile(r"^" + re.escape(base) + r"\.(\d+)$")
    found = []
    try:
        names = os.listdir(directory)
    except FileNotFoundError:
        names = []
    for name in names:
        m = part_re.match(name)
        if m and os.path.isfile(os.path.join(directory, name)):
            found.append((int(m.group(1)), os.path.join(directory, name)))
    if not found:
        raise FileNotFoundError("no parts found for %r" % arg)
    found.sort(key=lambda t: t[0])
    return [p for _, p in found]


class PartsReader:
    """A binary reader that streams the concatenation of parts in order."""

    def __init__(self, parts):
        self.parts = list(parts)
        self.idx = 0
        self.cur = None

    def _ensure(self):
        while self.cur is None:
            if self.idx >= len(self.parts):
                return False
            self.cur = open(self.parts[self.idx], "rb")
            self.idx += 1
        return True

    def read(self, n=-1):
        if n is None or n < 0:
            out = bytearray()
            while self._ensure():
                out += self.cur.read()
                self.cur.close()
                self.cur = None
            return bytes(out)
        out = bytearray()
        while len(out) < n:
            if not self._ensure():
                break
            chunk = self.cur.read(n - len(out))
            if not chunk:
                self.cur.close()
                self.cur = None
                continue
            out += chunk
        return bytes(out)

    def close(self):
        if self.cur is not None:
            self.cur.close()
            self.cur = None
